### PR TITLE
Adds c# simplifier to AutoRest.exe 

### PR DESCRIPTION
**Pull request by fearthecowboy**
AutoRest will now shorten references and use `using` statements in generated c# code by default.

Can be disabled with `-DontSimplify`

### CODE REVIEW

File: src/core/AutoRest.Core/Utilities/FileSystem.cs [hunk 1]
```diff
@@ -12,7 +12,7 @@ public class FileSystem : IFileSystem
     {
         public void WriteFile(string path, string contents)
         {
-            File.WriteAllText(path, contents, Encoding.UTF8);
+            File.WriteAllText(path, contents, new UTF8Encoding(false, true));
```
Comment (fearthecowboy) [inline src/core/AutoRest.Core/Utilities/FileSystem.cs]: Suppresses UTF-8 BOM in outputs

File: src/core/AutoRest/Simplify/CSharpSimplifier.cs [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+                return mscorlib;
+            }
+        }
+
+        public async Task Run()
+        {
+            var op = new AzureAsyncOperation();
+            var restOp = new RestException();
+
```
Comment (olydis) [inline src/core/AutoRest/Simplify/CSharpSimplifier.cs]: I assume those two guys make sure that `AppDomain.CurrentDomain.GetAssemblies()` contains everything we need?

File: src/core/AutoRest/Simplify/CSharpSimplifier.cs [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+            var files = Settings.Instance.FileSystem.GetFiles(Settings.Instance.OutputDirectory, "*.cs",
+                    SearchOption.AllDirectories).
+                ToDictionary(each => each, each => Settings.Instance.FileSystem.ReadFileAsText(each));
+
+            var projectId = ProjectId.CreateNewId();
+            var solution = new AdhocWorkspace().CurrentSolution
+                .AddProject(projectId, "MyProject", "MyProject", LanguageNames.CSharp)
+                .AddMetadataReference(projectId, Mscorlib)
+                .AddMetadataReference(projectId, AppDomain.CurrentDomain.GetAssemblies()
```
Comment (olydis) [inline src/core/AutoRest/Simplify/CSharpSimplifier.cs]: would `typeof(AzureAsyncOperation).Assembly` do the same? Same with `RestException` below

File: src/core/AutoRest/Simplify/AdHocEqualityComparer.cs [hunk 4]
```diff
@@ -0,0 +1,37 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+// 
+
+using System;
+using System.Collections.Generic;
+
+namespace AutoRest.Simplify
+{
+    public class AdHocEqualityComparer<T> : IEqualityComparer<T>
```
Comment (olydis) [inline src/core/AutoRest/Simplify/AdHocEqualityComparer.cs]: Think we added one of those today :D 
see `AutoRest.Core.Utilities.EqualityComparer`

File: src/core/AutoRest/Simplify/CSharpSimplifier.cs (hunk 3 above)
Comment (fearthecowboy) [inline src/core/AutoRest/Simplify/CSharpSimplifier.cs]: Will fix when moving to dotnet core for autorest.

**Pull request merged**